File: python/course-repo/exercise/04-concurrency-and-rate-limits/src/webhooks/worker.py
import asyncio
import logging
from concurrent.futures import ThreadPoolExecutor

from temporalio.client import Client
from temporalio.worker import Worker

from .activities import deliver_webhook
from .shared import TASK_QUEUE


# Show activity.logger.info(...) output; default WARNING would drop INFO.
logging.basicConfig(level=logging.INFO)


async def main() -> None:
    client = await Client.connect("localhost:7233")
    with ThreadPoolExecutor(10) as executor:
        worker = Worker(
            client,
            task_queue=TASK_QUEUE,
            activities=[deliver_webhook],
            activity_executor=executor,
            # TODO: cap how many activities this worker dispatches per second so
            # we don't overwhelm the downstream service. The Worker constructor
            # accepts `max_activities_per_second=<float>`. Try 2.0 first.
        )
        print(f"Worker running on task queue '{TASK_QUEUE}'")
        await worker.run()


if __name__ == "__main__":
    asyncio.run(main())

File: java/course-repo/exercise/04-concurrency-and-rate-limits/src/main/java/webhooks/Worker.java
package webhooks;

import io.temporal.client.WorkflowClient;
import io.temporal.serviceclient.WorkflowServiceStubs;
import io.temporal.worker.WorkerFactory;
import io.temporal.worker.WorkerOptions;

import java.util.concurrent.CountDownLatch;

/**
 * The Worker program. It polls the Task Queue for Activity tasks and runs them.
 * A Standalone Activity is served by the same kind of Worker you would use for
 * Workflows.
 */
public class Worker {
  public static void main(String[] args) throws InterruptedException {
    WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
    WorkflowClient client = WorkflowClient.newInstance(service);
    WorkerFactory factory = WorkerFactory.newInstance(client);

    WorkerOptions workerOptions =
        WorkerOptions.newBuilder()
            .setMaxConcurrentActivityExecutionSize(10)
            // TODO: cap how many Activities this Worker dispatches per second so
            // we don't overwhelm the downstream service.
            // The WorkerOptions method is: setMaxTaskQueueActivitiesPerSecond(<number>)
            // Try 2 first to match the receiver's 2 req/sec rate limit.
            .build();

    io.temporal.worker.Worker worker = factory.newWorker(Shared.TASK_QUEUE, workerOptions);
    worker.registerActivitiesImplementations(new WebhookActivitiesImpl());

    factory.start();
    System.out.println("Worker running on task queue '" + Shared.TASK_QUEUE + "'");

    // Poller threads run in the background; block the main thread so the JVM stays up.
    new CountDownLatch(1).await();
  }
}
